playbooks/openshift-node/private/join.yml

```diff
@@ -0,0 +1,87 @@
+---
+- name: Evaluate bootstrapped nodes
+  hosts: localhost
+  gather_facts: no
+  connection: local
+  tasks:
+  - name: Add all nodes that are bootstrapped
+    add_host:
+      name: "{{ item }}"
+      groups: oo_nodes_to_bootstrap
+      ansible_ssh_user: "{{ g_ssh_user | default(omit) }}"
+      ansible_become: "{{ g_sudo | default(omit) }}"
+    with_items: "{{ groups.oo_nodes_to_config | default([]) }}"
+    when:
+    - hostvars[item].openshift is defined
+    - hostvars[item].openshift.common is defined
+    - hostvars[item].openshift_is_bootstrapped | bool
+    changed_when: False
+
+- name: Gather facts from nodes to bootstrap
+  hosts: oo_nodes_to_bootstrap
+  gather_facts: no
+  tasks:
+  - name: Gather node information
+    openshift_facts:
+      role: node
+      local_facts:
+        bootstrapped: true
+
+  - name: Copy master bootstrap config locally
+    slurp:
+      src: "/etc/origin/master/bootstrap.kubeconfig"
+    delegate_to: "{{ groups.oo_first_master.0 }}"
+    register: bootstrap_credentials
+
+  - name: Distribute bootstrap kubeconfig if one does not exist
+    copy:
+      content: "{{ bootstrap_credentials.content | b64decode }}"
+      dest: "/etc/origin/node/bootstrap.kubeconfig"
+      mode: 0700
+      owner: root
+      group: root
+      force: no
+
+- name: Approve any pending CSR requests from inventory nodes
+  hosts: oo_first_master
+  gather_facts: no
+  tasks:
+  - name: Dump all candidate bootstrap hostnames
+    debug:
+      msg: "{{ groups['oo_nodes_to_bootstrap'] | default([]) }}"
+
+  - name: Find all node names
+    set_fact:
+      l_nodes_to_join: "{{ groups['oo_nodes_to_bootstrap'] | default([]) | map('extract', hostvars) | map(attribute='openshift.node.nodename') | list }}"
+
+  - name: Dump the bootstrap hostnames
+    debug:
+      msg: "{{ l_nodes_to_join }}"
+
+  - name: Approve bootstrap nodes
```

Comment:
Why would we need to approve CSRs here? This is what `bootstrap-autoapprover` is for.

I've hit this when an installation with several nodes is being setup - services on additional nodes are being prepared, but not started. As a result, this task would hang until times out (see Fedora Atomic results)

Reply:
auto approver is a security decision and so it defaults to off.  When auto approver is off, certs have very long expirations.  when auto approver is on we set aggressive expirations.

Reply:
When autoapprover is off `oc_adm_csr` needs to happen while `origin-node` is starting - otherwise either of those would time out. That would require the code to be async, that would probably be too complex

Reply:
CSRs shouldn't time out.